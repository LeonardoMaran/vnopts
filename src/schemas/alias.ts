import { Schema, SchemaParameters } from '../schema';
import { RedirectResult, Utils, ValidateResult } from '../types';

interface AliasSchemaParameters extends SchemaParameters<any> {
  sourceName: string;
}

export class AliasSchema extends Schema<any, AliasSchemaParameters> {
  private _sourceName: string;

  constructor(parameters: AliasSchemaParameters) {
    super(parameters);
    this._sourceName = parameters.sourceName;
  }

  public expected(utils: Utils) {
    return utils.schemas[this._sourceName].expected(utils);
  }

  public validate(value: unknown, utils: Utils): ValidateResult {
    return utils.schemas[this._sourceName].validate(value, utils);
  }

  public redirect(
    _value: this['_valueType'],
    _utils: Utils,
  ): RedirectResult<this['_valueType']> {
    return this._sourceName;
  }
}
