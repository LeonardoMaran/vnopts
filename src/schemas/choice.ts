import { Schema, SchemaHandlers, SchemaParameters } from '../schema';
import {
  DeprecatedResult,
  ForwardResult,
  RedirectResult,
  TransferTo,
  Utils,
  ValidateResult,
} from '../types';
import { comparePrimitive, mapFromArray } from '../utils';

interface ChoiceInfo {
  value: ChoiceValue;
  deprecated?: boolean;
  redirect?: TransferTo;
  forward?: TransferTo;
}

type ChoiceValue = undefined | null | boolean | number | string;

interface ChoiceSchemaParameters
  extends SchemaParameters<ChoiceValue>,
    SchemaHandlers<ChoiceValue> {
  choices: Array<ChoiceValue | ChoiceInfo>;
}

export class ChoiceSchema extends Schema<ChoiceValue, ChoiceSchemaParameters> {
  private _choices: Map<ChoiceValue, ChoiceInfo>;

  constructor(parameters: ChoiceSchemaParameters) {
    super(parameters);
    this._choices = mapFromArray(
      parameters.choices.map(
        choice =>
          choice && typeof choice === 'object' ? choice : { value: choice },
      ),
      'value',
    );
  }

  public expected({ descriptor }: Utils) {
    const choiceValues = Array.from(this._choices.keys())
      .map(value => this._choices.get(value)!)
      .filter(choiceInfo => !choiceInfo.deprecated)
      .map(choiceInfo => choiceInfo.value)
      .sort(comparePrimitive)
      .map(descriptor.value);

    const head = choiceValues.slice(0, -2);
    const tail = choiceValues.slice(-2);
    return head.concat(tail.join(' or ')).join(', ');
  }

  public validate(value: unknown): ValidateResult {
    return this._choices.has(value as any);
  }

  public deprecated(value: ChoiceValue): DeprecatedResult<ChoiceValue> {
    const choiceInfo = this._choices.get(value);
    return choiceInfo && choiceInfo.deprecated ? { value } : false;
  }

  public forward(value: ChoiceValue): ForwardResult<ChoiceValue> {
    const choiceInfo = this._choices.get(value);
    return choiceInfo ? choiceInfo.forward : undefined;
  }

  public redirect(value: ChoiceValue): RedirectResult<ChoiceValue> {
    const choiceInfo = this._choices.get(value);
    return choiceInfo ? choiceInfo.redirect : undefined;
  }
}
