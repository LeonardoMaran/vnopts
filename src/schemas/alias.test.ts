import { eachHandler } from '../../tests/__helpers__/utils';
import { AliasSchema } from './alias';
import { ChoiceSchema } from './choice';

const name = '<key>';
const validValue = '<valid-value>';
const invalidValue = '<invalid-value>';

const sourceName = '<source-key>';

describe.each`
  parameters | input                       | output                          | hasWarnings
  ${{}}      | ${{ [name]: invalidValue }} | ${Error}                        | ${false}
  ${{}}      | ${{ [name]: validValue }}   | ${{ [sourceName]: validValue }} | ${false}
`(
  '',
  eachHandler<AliasSchema>(
    AliasSchema,
    { name, sourceName },
    {
      schemas: [
        ChoiceSchema.create<ChoiceSchema>({
          name: sourceName,
          choices: [validValue],
        }),
      ],
    },
  ),
);
