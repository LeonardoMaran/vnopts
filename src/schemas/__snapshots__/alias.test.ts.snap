// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema {"name":"<key>","sourceName":"<source-key>"} with input {"<key>":"<invalid-value>"} 1`] = `"Invalid \\"<key>\\" value. Expected \\"<valid-value>\\", but received \\"<invalid-value>\\"."`;
